linera-faucet/server/src/lib.rs

```diff
@@ -118,17 +123,57 @@ where
     C: ClientContext,
 {
     async fn do_claim(&self, owner: AccountOwner) -> Result<ClaimOutcome, Error> {
-        let client = self.context.lock().await.make_chain_client(self.chain_id)?;
+        let main_client = self
+            .context
+            .lock()
+            .await
+            .make_chain_client(self.main_chain_id)?;
+        let maybe_tmp_chain_id = *self.tmp_chain_id.lock().await;
+        let main_balance = main_client.local_balance().await?;
+        let tmp_chain_id = match maybe_tmp_chain_id {
+            Some(tmp_chain_id) => tmp_chain_id,
+            None => {
+                let key_pair = main_client.key_pair().await?;
+                let balance = self
+                    .amount
+                    .try_add(MAX_FEE)?
+                    .try_mul(u128::from(self.max_claims_per_chain))?
+                    .try_add(MAX_FEE)? // One more block fee for closing the chain.
+                    .min(main_balance.try_sub(MAX_FEE)?);
```

Comment:
This could be extracted to a method (`chain_init_balance`) or even the field on the struct. We don't use `max_claims_per_chain` except for calculating this. It could be an argument to the constructor but the constructor could do the calculation on struct instantiation.